core/src/main/java/io/grpc/NameResolver.java

```diff
@@ -0,0 +1,109 @@
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+import java.util.concurrent.Executor;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A pluggable component that resolves a target URI (which is broken down into 3 parts as described
+ * below) and return addresses to the caller.
+ *
+ * <p>The format of the target URI is {@code
+ * "[<scheme>://[<name-system-authority>]/]<endpoint_name>"}.
+ * <ul>
+ *   <li>The {@code scheme} indicates the name-system to be used.  Example {@code scheme}s to be
+ *   supported include {@code dns}, {@code zookeeper} and {@code etcd}.
+ *   <li>The {@code name-system-authority} indicates scheme-specific bootstrap information, e.g.,
+ *   the {@code IP:port} of the DNS server or ZooKeeper server to use.
+ *   <li>The {@code endpoint_name} indicates a concrete name to be looked up in a given name-system
+ *   identified by the scheme and the authority.
+ * </ul>
+ *
+ * <p>{@code NameResolver} has no knowledge of load-balancing. The addresses of a target may be
+ * changed over time, thus the caller registers a {@link Listener} to receive continuous updates.
+ */
+@ExperimentalApi
+public interface NameResolver {
+  /**
+   * Creates handle for resolving an endpoint name in a name-system.
+   *
+   * @param scheme indicates the name-system to be used.
+   * @param nsAuthority indicates scheme-specific bootstrap information, typically the address of
+   *                    the name server.
+   * @param endpoint the concrete name to be looked up in the given name-system.
+   */
+  Handle resolve(String scheme, @Nullable String nsAuthority, String endpoint);
```

Comment:
Let's swap this to just a plan URI. That prevents us from needing to define the semantics of the various portions, which is actually quite hard to do. It's also my understanding that that aligns us with C core.